airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGenerator.kt

```diff
@@ -109,18 +113,123 @@ class SnowflakeSqlGenerator(private val retentionPeriodDays: Int) : SqlGenerator
+                .joinToString(" AND \n") {
+                    """
+                    |  ($targetTable.$it=$sourceTable.$it
+                    |    OR ($targetTable.$it IS NULL AND $sourceTable.$it IS NULL))
+                    |""".trimMargin()
+                }
+        // First, compare the cursors.
+        // Then, break ties with extracted_at. (also explicitly check for both source_table and
+        // target_table
+        // having null cursor because NULL != NULL in SQL)
+        //    - Thanks to edgao
+        //    (author referred to this comment for writing code so might as well put it here)
+        val cursorComparisonCondition =
+            stream.cursor
+                .map { it.name.quoted() }
+                .map {
+                    """
+                        |($targetTable.$it < $sourceTable.$it
+                        |  OR ($targetTable.$it = $sourceTable.$it AND $targetAbExtractedAt < $sourceTable.$abExtractedAt)
+                        |  OR ($targetTable.$it IS NULL AND $sourceTable.$it IS NULL AND $targetAbExtractedAt < $sourceTable.$abExtractedAt)
+                        |  OR ($targetTable.$it IS NULL AND $sourceTable.$it IS NOT NULL)
+                        |)
+                    """.trimMargin()
+                }
+                .orElse("$targetAbExtractedAt < $sourceTable.$abExtractedAt")
+        val whenMatchedCdcDeleteCondition =
+            if (stream.columns.containsKey(cdcDeletedAtColumn))
+                "\n WHEN MATCHED AND $sourceTable.${cdcDeletedAtColumn.name.quoted()} IS NOT NULL AND $cursorComparisonCondition THEN DELETE"
+            else ""
+        val whenNotMatchedCdcSkipCondition =
+            if (stream.columns.containsKey(cdcDeletedAtColumn))
+                "AND $sourceTable.${cdcDeletedAtColumn.name.quoted()} IS NULL"
+            else ""
+        val columnsForUpdateList =
+            stream.columns.keys.map { it.name.quoted() } +
+                listOf(
+                    JavaBaseConstants.COLUMN_NAME_AB_RAW_ID.uppercase().quoted(),
+                    JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT.uppercase().quoted(),
+                    JavaBaseConstants.COLUMN_NAME_AB_GENERATION_ID.uppercase().quoted()
+                )
+        val abMetaColumn = JavaBaseConstants.COLUMN_NAME_AB_META.uppercase().quoted()
+
+        val upsertSql =
+            """
+            |MERGE INTO ${stream.id.quotedFinalTableId(finalSuffix)} as $targetTable
+            |USING (
+            |${selectTypedRecordsFromRawTable(stream, minRawTimestamp, safeCast).replaceIndent("  ")}
+            |) $sourceTable
+            |ON
+            |${pkEqualityMatch.replaceIndent("  ")} $whenMatchedCdcDeleteCondition
+            |WHEN MATCHED AND $cursorComparisonCondition THEN UPDATE SET
+            |${columnsForUpdateList.joinToString(", \n") { "$it = $sourceTable.$it" }.replaceIndent("  ")},
+            |  $abMetaColumn = $sourceTable.$abMetaColumn
```

Comment:
doublechecking: this works because `sourceTable` already includes all the casting errors?

Reply:
yeah. aliased with the same `AB_META` string